### PR TITLE
feat(hydroflow_datalog): add syntax for "splatting" columns

### PR DESCRIPTION
**Pull request by shadaj**
feat(hydroflow_datalog): add syntax for "splatting" columns

The * operator can be used on the left-hand side of a rule to indicate that each row should be duplicated into many rows with that column being replaced by every value in iterable corresponding to the original expression.

### CODE REVIEW

**Comment by MingweiSamuel** [comment]
Would prefer to use ellipses for splitting `...` which would match the variadics crate and also prevent conflict with Deref

File: hydroflow_datalog_core/src/lib.rs [hunk 1]
```diff
@@ -665,64 +715,41 @@ fn apply_aggregations(
...
-                after_group_lookups.push(parse_quote_spanned!(get_span(field.span)=> i));
-            }
-        }
-    }
-
     let fold_keyed_input_type =
         repeat_tuple::<syn::Type, syn::Type>(|| parse_quote!(_), fold_keyed_exprs.len());
 
+    // TODO(shadaj): use splat indices
+
```
Comment (MingweiSamuel) [inline hydroflow_datalog_core/src/lib.rs:722]: ? still needed?

**Review by MingweiSamuel — APPROVED** [review]

**Review by davidchuyaya — APPROVED** [review]
lgtm?

**Comment by shadaj** [comment]
> Would prefer to use ellipses for splatting `...` which would match the variadics crate and also prevent conflict with Deref (if we ever allow full expressions)
> 
> 
> 
> (And also cause I have a beef with Python in general)

Ok so the reason I used * is because ... somewhat implies the semantics of generating a dynamic number of columns, which this is not. Instead we dynamically expand rows into many rows.

**Comment by MingweiSamuel** [comment]
Oh, so this is actually a sort of cartesian product?

**Comment by MingweiSamuel** [comment]
Ah, flatten really

**Pull request merged**